Overview docs for MLflow Pipelines (#6145)
mlflow/mlflow: docs/source/pipelines.rst

.. _pipelines:

===============================
MLflow Pipelines (experimental)
===============================

MLflow Pipelines is an opinionated framework for structuring MLOps workflows that simplifies and
standardizes machine learning application development and productionization. MLflow Pipelines
makes it easy for data scientists to follow best practices for creating production-ready ML

> deliverables, allowing them to focus on utilizing their expert knowledge to develop excellent
> models. MLflow Pipelines also enables ML engineers and DevOps teams to seamlessly deploy these

Reviewer: ```suggestion
deliverables, allowing them to focus on developing excellent
models. MLflow Pipelines also enables ML engineers and DevOps teams to seamlessly deploy these
```

Reviewer: Not sure if the prior was relevant

Author: Agreed, it wasn't. Thanks for the simplification!

models to production and incorporate them into applications.

> MLflow Pipelines provides predefined production-quality pipeline templates for common ML problem

Reviewer: ```suggestion
MLflow Pipelines provides production-quality pipeline templates for common ML problem
```

Reviewer: Doesn't "template" imply predefined?

Author: Indeed, it does! Fixed :)

<steps-key-concept>`, ensuring that steps are only executed when associated :py:ref:`code or
configurations <pipeline-repositories-key-concept>` have changed. This enables data scientists,
ML engineers, and DevOps teams to iterate very quickly within their domains of expertise. MLflow
offers :ref:`APIs <pipelines-apis>` and a :ref:`CLI <cli>` for executing pipelines and inspecting
their results.

Installation
------------
MLflow Pipelines is available as an extension of the
`MLflow Python library <https://pypi.org/project/mlflow/>`_. You can install MLflow Pipelines
from PyPI as follows:

.. code-block:: sh

  pip install mlflow[pipelines]

.. note::
  You can install MLflow Pipelines from a Databricks Notebook by running
  ``%pip install mlflow[pipelines]`` or install MLflow Pipelines on a Databricks Cluster
  by following the instructions at
  https://docs.databricks.com/libraries/cluster-libraries.html#install-a-library-on-a-cluster.

  Note that `Databricks Runtime <https://docs.databricks.com/runtime/dbr.html>`_ version 11.0
  or greater is required in order to install MLflow Pipelines on Databricks.

Key Concepts
------------

.. _steps-key-concept:

Steps
~~~~~

> Steps represent individual modeling or MLOps procedures, such as ingesting data,

Reviewer: ```suggestion
Steps represent individual MLOps procedures, such as ingesting data,
```

Author: MLOps is distinct from model development, so I'd prefer to keep `modeling` in there:

```
MLOps or ML Ops is a set of practices that aims to deploy and maintain machine learning models in production reliably and efficiently. The word is a compound of "machine learning" and the continuous development practice of DevOps in the software field.
```

https://en.wikipedia.org/wiki/MLOps

Reviewer: TIL again lol

      import os
      from mlflow.pipelines import Pipeline
      from mlflow.pyfunc import PyFuncModel

      os.chdir("~/mlp-regression-template")
      regression_pipeline = Pipeline(profile="local")
      # Run the full pipeline
      regression_pipeline.run()
      # Inspect the model training results
      regression_pipeline.inspect(step="train")
      # Load the trained model
      regression_model_pipeline: PyFuncModel = regression_pipeline.get_artifact("model")

    .. code-block:: sh

      git clone https://github.com/mlflow/mlp-regression-template
      cd mlp-regression-template
      # Run the full pipeline
      mlflow pipelines run --profile local
      # Inspect the model training results
      mlflow pipelines inspect --step train --profile local
      # Inspect the resulting model performance evaluations
      mlflow pipelines inspect --step evaluate --profile local

.. _pipeline-repositories-key-concept:

Pipeline Repositories
#####################

Pipelines are structured as git repositories with a standardized directory hierarchy. The following
example pipeline repository layout adapted from the |MLflow Regression Pipeline repository|

> showcases the main components that are common across all pipeline repositories.

Reviewer: ```suggestion
shows the main components that are common across all pipeline repositories.
```

Reviewer: Or illustrates

Author: I think the issue here was less the word choice and more how long / wordy the sentence was. Broke this up / changed to

```
Pipelines are structured as git repositories with a standardized directory hierarchy. The following
example provides an overview of the pipeline repository structure.  It is adapted from the
|MLflow Regression Pipeline repository|.
```

Lmk what you think :)

Reviewer: Amazing!

      split:
        split_ratios: {{SPLIT_RATIOS|default([0.75, 0.125, 0.125])}}
      transform:
        transformer_method: steps.transform.transformer_fn
      train:
        estimator_method: steps.train.estimator_fn
      evaluate:
        validation_criteria:
          - metric: root_mean_squared_error
            threshold: 10
          - metric: weighted_mean_squared_error
            threshold: 20
      register:
        model_name: "taxi_fare_regressor"
    metrics:
      custom:
        - name: weighted_mean_squared_error
          function: weighted_mean_squared_error
          greater_is_better: False
      primary: "root_mean_squared_error"

.. _profiles-key-concept:

Profiles
~~~~~~~~

As you develop and iterate on pipelines, you will likely need to explore a variety of configuration
and code changes to arrive at an optimal solution. For example, in the context of model training,
it is often important to experiment with different datasets, feature transformations, and model
parameters. Pipeline Profiles offer a systematic approach for incrementally testing changes in
collaborative contexts, enabling multiple team members to collaborate on the same pipeline

> steps while minimizing code and configuration conflicts. Profiles also make it very easy to

Author: I think the existing phrasing makes the value prop clearer.

        pip install dbx
        # Clone the MLflow Regression Pipeline repository
        git clone https://github.com/mlflow/mlp-regression-template
        # Enter the MLflow Regression Pipeline repository directory and configure dbx within it
        cd mlp-regression-template
        dbx configure
        # Use dbx to enable syncing from the repository directory to Databricks Repos
        dbx sync repo -d mlp-regression-template
        # Iteratively make changes to files in the repository directory and observe that they
        # are automatically synced to Databricks Repos
        ...

4. Test changes by running the pipeline and observing the results it produces. MLflow Pipelines
   intelligently caches results from each :ref:`Pipeline Step <steps-key-concept>`, ensuring that
   steps are only executed if their inputs, code, or configurations have changed, or if such
   changes have occurred in dependent steps. Once you are satisfied with the results of
   your changes, commit them to a branch of the :ref:`Pipeline Repository
   <pipeline-repositories-key-concept>` in order to ensure reproducibility, and share or review the

>    changes with your team.

Author: I think we need to offer some explanation. git isn't as commonly used in the DS community.

>    changes with your team.

Reviewer: TIL!
